### DISCUSSION TITLE
Renovate config to update Gemfile.lock only

### DISCUSSION THREAD

**Discussion by layerssss** [renovatebot/renovate#8673] [Request Help]
Hi, I'm already using the pretty much the default config for a create-react-app js (with Yarn) project for a while. I understand it needs to pin versions in `package.json` but it works well. Now I'm looking into configuring it for a Rails project.

Will I able to achive the below behaviour?

* My `Gemfile` lists pretty much all gems without version/range, e.g: `gem 'sidekiq'`, except one or two having a range to prevent it goes beyond current minor version, e.g `gem 'rails', '~6.0.0'`
* My `Gemfile.lock` goes with source control and it locks version of every gems listed in the `Gemfile` plus versions of *their dependencies*
* I want to renovate to update items in the `Gemfile.lock` if new versions of those are availble, but not changing the content of `Gemfile`.

**Comment by viceice** [comment]
Lock file only update should be possible with lockfile maintenance. This will update all transitive deps to their latest version.

**Comment by clemens** [comment]
This would be a great change:

- It's an absolutely realistic use case for various packages where people tend to want to stay at the latest version (e.g. `bundler-audit`) or for packages with a high release frequency (e.g. `rubocop`).
- It would arguably also educate people to start to care more about pinning versions: If renovate keeps opening MRs where people don't quite understand why they're created, they'll hopefully get educated.

One thing that would make such a change smooth(er) would be to include a little note in the pull request message. I'm thinking of something like "Note: This dependency has no version constraint and will thus always receive pull request with the latest version. Consider pinning your version if this behavior is undesirable." (and "pinning your version" could be linked to an appropriate tutorial, e.g. https://guides.rubygems.org/patterns/#pessimistic-version-constraint).

**Reply by rarkins** [reply]
Is there anything technically about Ruby/Rubygems which leads people to use "any version" constraints instead of at least using a major range, or is it just an early convention which people have adopted without thinking much about it? For example in Node.js it would be regarded terrible practice to use `*` for dependencies and instead it defaults to e.g. `^1.0.0` so that you wouldn't get a breaking 2.x version without changes to your Gemfile.

**Reply by layerssss** [reply]
@rarkins It might be somewhat historical as `bundler` has the `Gemfile.lock` since its inception, so people always use it and it has been reliable without having to specify constraints in `Gemfile`. (Gemfile.lock is always version controlled and enforced, not just by bundler CLI, but also by Rails).

It wasn't the story for Node.js though. `npm-shrinkwrap`, `package-lock`, and `yarn.lock` came much later to the party and had not been that stable until many years later.

**Reply by clemens** [reply]
There's multiple factors at play here:

- Ruby has never, in any way, enforced version constraints for dependencies. Even in the old (pre-Bundler) days, you could simply do `config.gem 'your-dependency'` and all was well. This is probably also the result from Ruby using a DSL here instead of a typical key-value format like JSON. But either way, the fact that Renovate expects this is unexpected to a Ruby developer.
- The general approach (that I've seen used in dozens of projects that I've worked on) seems to be to set strict constraints for some dependencies (e.g. Rails, where even upgrading to the next minor version is sometimes quite a hassle) and lenient or even no constraints at all for other dependencies (e.g. tools for linting/auditing, test coverage reporting etc. where you don't expect failures until you integrate very deeply with them).
- The [documentation](https://docs.renovatebot.com/modules/manager/bundler/) doesn't indicate that the field is mandatory in order to manage the dependency version using Renovate.
- The output of running Renovate doesn't tell me (not even with the highest log level) that certain packages have been skipped due to missing version constraints.

Minor nitpick on the side that might be specific to my personal notion: When I don't specify a constraint, my mind is basically in "I don't care about the version as long as the package is there" mode. On the other hand defaulted to suggesting an upgrade to Postgres 13 (which had locked to `12.1` because I actually cared about this linked in `Dockerfile`/`docker-compose.yml` and `.gitlab-ci.yml`). So this was confusing on 2 fronts to me: I'd have expected Renovate to suggest upgrades to Ruby gems (because they had no version constraints) but wouldn't have expected it to suggest Postgres upgrades (or, if at all, suggest `12.6` which at least stays in the same major release).

Steps that could be taken to remedy the situation:

- Update documentation to reflect the behavior. This should be quite easy and I'm happy to provide a fix, if desired.
- Show helpful output (i.e. which packages were ignored and why) when Renovate runs against a `Gemfile`

Separately of this, a discussion regarding expected behavior would IMO be warranted.

**Reply by rarkins** [reply]
Docs: PR appreciated
Output: Does the `packageFiles with update` debug message list dependency as well as a `skipReason`?

**Reply by clemens** [reply]
Doc update here: https://github.com/renovatebot/renovate/pull/11644

As for your question: Yes, the `skipReason` shows up (see screenshot below). One could argue that the given reason is rather ambiguous: "no-version" could also mean that there is no version for Renovate to upgrade to (I could be on the latest version or on the latest version that satisfies my given version constraint). The meaning only becomes evident once you investigate the source code or deduce from other output (as seen in the 2nd screenshot below – apparently the presence/absence of `currentValue` indicates that a version constraint is present). Would you be open to change this to something less ambiguous like "missing-version-constraint"?

![Screenshot 2021-09-08 at 15 24 08](https://user-images.githubusercontent.com/3749/132517666-6a7250dd-db32-4764-9b3c-09e3a19ad58c.png)

![Screenshot 2021-09-08 at 15 27 20](https://user-images.githubusercontent.com/3749/132518149-02526fa0-4d0e-4371-baa9-1882740815d0.png)

**Reply by igas** [reply]
I guess another reason not to have constraints in gemfile, is due to rubygems not having nested dependencies as npm does, it's in a way similar to `yarn --flatten`. So if I specify a dependency version of some unimportant gem, it might prevent the important gem from upgrading silently.

In addition, I have not checked previous lockfile maintenance suggestions yet, but it would be also great to have not defined dependencies to be upgraded too. Say I use both `octokit` and `sentry-raven`, both relying on `faraday`. I never use `faraday` in my code, so it's not my dependency, but I do care if this HTTP library has security vulnerabilities that I get an update. This is how dependabot does it, unfortunately, dependabot does not support groups, so we use both.

We'd be really happy to move fully from dependabot, but those 2 issues are no go for now.

**Reply by rarkins** [reply]
No objection from our side, #7883 is marked as `status:ready` so simply awaiting implementation by someone
